numpy/_core/defchararray.py

```diff
@@ -56,17 +56,18 @@ def _is_unicode(arr):
     return False
 
 
-def _to_bytes_or_str_array(result, output_dtype_like=None):
+def _to_bytes_or_str_array(result, output_dtype_like):
     """
     Helper function to cast a result back into an array
     with the appropriate dtype if an object array must be used
     as an intermediary.
     """
+    if result.size == 0:
+        # Calling asarray & tolist in an empty array would result
+        # in a float64 array
+        return result.astype(output_dtype_like.dtype)
     ret = numpy.asarray(result.tolist())
-    dtype = getattr(output_dtype_like, 'dtype', None)
-    if dtype is not None:
-        return ret.astype(type(dtype)(_get_num_chars(ret)), copy=False)
-    return ret
+    return ret.astype(type(output_dtype_like.dtype)(_get_num_chars(ret)))
```

Comment:
This looks unnecessary.  Just use `numpy.asarray(result.tolist(), dtype=type(output_dtype_like.dtype))`.

You shouldn't even need the `tolist()`, making:
```
ret.astype(type(dtype), copy=False)
```
possible.  In theory limits user dtypes a bit, so I am happy to just keep the tolist for now.

Reply:
Ah right, for 0 size tolist loses the shape info, sorry.  Maybe the comment is slightly off, the reason is (at least mainly?) the loss of shape info, not the dtype.

OK, so we still need the tolist, because the result has an extra dimension (once upon a time it was more critical); that means the 0 size return is missing that extra dimensions, but...
I think you are right that unless you were to modify the C-code to add that dimension, you need that dance then.

I guess at this point it is about how *much* of this mess we want to clear out in this iteration, and I will leave that up to you.  Fine to just go with this as a step in the right direction.

Reply:
I'm okay with just doing this small step, since most of this is going in numpy 2.0 either way.